### PR TITLE
New case of convert_to_virtual_disk

### PR DESCRIPTION
**Pull request by hellohellenmao**
1. Create input and output images.
2. Setup loop device with the output image.
3. Convert the input image to the loop device.

ID: 1788805
Signed-off-by: Tingting Mao <timao@redhat.com>

### CODE REVIEW

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.convert_to_virtual_disk: PASS (8.54 s)

**Comment by hellohellenmao** [comment]
@lolyu @YongxueHong Help to take a review, please.

File: qemu/tests/convert_to_virtual_disk.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from avocado.utils import process
+
+from virttest import data_dir
+from virttest.qemu_storage import QemuImg
+
+
+def run(test, params, env):
+    """
+    Convert a image to a virtual block device.
+    1. Create input and output images.
```
Comment (lolyu) [inline qemu/tests/convert_to_virtual_disk.py]: `source` and `dest` will be more appropriate.

File: qemu/tests/convert_to_virtual_disk.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    3. Convert the input image to the loop device.
+    :param test: Qemu test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+    input_img = params["images"].split()[0]
+    output_img = params["images"].split()[-1]
+    root_dir = data_dir.get_data_dir()
+    source = QemuImg(params.object_params(input_img), root_dir, input_img)
+    output = QemuImg(params.object_params(output_img), root_dir, output_img)
```
Comment (lolyu) [inline qemu/tests/convert_to_virtual_disk.py]: same as above.

File: qemu/tests/convert_to_virtual_disk.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    output_filename = output.image_filename
+
+    logging.info("Create the test image files.")
+    source.create(source.params)
+    output.create(output.params)
+
+    logging.info("Setup target loop device via 'losetup'.")
+    try:
```
Comment (lolyu) [inline qemu/tests/convert_to_virtual_disk.py]: please wrap the process of loose device setup, probe, and cleanup into functions.

File: qemu/tests/convert_to_virtual_disk.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                                   shell=True).decode()
+    params["image_name_target"] = target
+
+    logging.debug("Convert from %s to %s with cache mode none.",
+                  source.image_filename, output_filename)
+    fail_on((process.CmdError,))(source.convert)(
+        params.object_params(input_img), root_dir, cache_mode="none")
+
+    logging.info("Clean the loop device.")
+    process.run("losetup -d %s" % target)
```
Comment (YongxueHong) [inline qemu/tests/convert_to_virtual_disk.py]: I assume that do not `process.run("losetup -d %s" % target)` if raise errors.

File: qemu/tests/convert_to_virtual_disk.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    output.create(output.params)
+
+    logging.info("Setup target loop device via 'losetup'.")
+    try:
+        process.run("losetup -f %s" % output_filename)
+    except process.CmdError as err:
+        test.error("Command line failed: %s" % err)
+
+    target = process.system_output("losetup -l | tail -n1 | awk '{print $1}'",
+                                   shell=True).decode()
```
Comment (lolyu) [inline qemu/tests/convert_to_virtual_disk.py]: It cannot be guaranteed the lastest loop device is the one created, better to search by the back-file.

File: qemu/tests/cfg/convert_to_virtual_disk.cfg [hunk 6]
```diff
@@ -0,0 +1,15 @@
+- convert_to_virtual_disk:
+    only raw
+    virt_test_type = qemu
+    type = convert_to_virtual_disk
+    start_vm = no
+    create_image = yes
+    remove_image = yes
+    images = "input output"
+    image_name_input = "images/input"
+    image_name_output = "images/output"
```
Comment (lolyu) [inline qemu/tests/cfg/convert_to_virtual_disk.cfg]: `source` && `dest`

File: qemu/tests/convert_to_virtual_disk.py (hunk 5 above)
Comment (lolyu) [reply qemu/tests/convert_to_virtual_disk.py]: Please wrap the functionality to probe loop-device by back-file into a function.

**Comment by lolyu** [comment]
Please modify your commit log following this [blog](https://chris.beams.io/posts/git-commit/).

File: qemu/tests/convert_to_virtual_disk.py (hunk 4 above)
Comment (hellohellenmao) [reply qemu/tests/convert_to_virtual_disk.py]: Updated, thanks.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.convert_to_virtual_disk: PASS (9.11 s)

File: provider/qemu_img_utils.py [hunk 7]
```diff
@@ -115,3 +115,32 @@ def check_flag(strace_log, target_file, flag):
         for line in lines:
             logging.debug(line.strip())
         return any(flag in line for line in lines)
+
+
+def setup_loop_dev(export_img):
```
Comment (lolyu) [inline provider/qemu_img_utils.py]: this could simply be like:
```python
def setup_loop_dev(image_path):
    cmd_result = process.run("losetup -f %s --show" % image_path)
    return cmd_result.stdout_text.strip()
```

File: provider/qemu_img_utils.py [hunk 8]
```diff
@@ -115,3 +115,32 @@ def check_flag(strace_log, target_file, flag):
...
+    return target
+
+
+def free_loop_dev(loop_dev):
+    """
+    Free a loop device.
+
+    :param loop_dev: The loop device will be free
+    """
+    try:
```
Comment (lolyu) [inline provider/qemu_img_utils.py]: same as above, no need to raise a `avocado.TestError`

Comment (hellohellenmao) [reply provider/qemu_img_utils.py]: Updated, thanks.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.convert_to_virtual_disk: PASS (8.74 s)

File: qemu/tests/convert_to_virtual_disk.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    target = setup_loop_dev(mid_filename)
+    params["image_name_target"] = target
+
+    logging.debug("Convert from %s to %s with cache mode none.",
+                  source.image_filename, mid_filename)
+    fail_on((process.CmdError,))(source.convert)(
+        params.object_params(src_img), root_dir, cache_mode="none")
+
+    logging.info("Clean the loop device.")
+    free_loop_dev(target)
```
Comment (YongxueHong) [inline qemu/tests/convert_to_virtual_disk.py]: `free_loop_dev(target)` maybe not be executed if test failed.

File: provider/qemu_img_utils.py [hunk 10]
```diff
@@ -115,3 +115,23 @@ def check_flag(strace_log, target_file, flag):
...
+    return cmd_result.stdout_text.strip()
+
+
+def free_loop_dev(loop_dev):
+    """
+    Free a loop device.
+
+    :param loop_dev: The loop device will be free
+    """
+    process.run("losetup -d %s" % loop_dev)
```
Comment (YongxueHong) [inline provider/qemu_img_utils.py]: Please don't add functions that are not related to `qemu-img`. 
Thanks.

File: qemu/tests/convert_to_virtual_disk.py (hunk 9 above)
Comment (hellohellenmao) [reply qemu/tests/convert_to_virtual_disk.py]: Updated, thanks.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.convert_to_virtual_disk: PASS (8.83 s)
